logstash-core/src/test/java/org/logstash/EventTest.java

```diff
@@ -516,4 +516,53 @@ public void removeMetadata() {
         assertTrue(event.getMetadata().isEmpty());
         assertFalse(event.includes("[@metadata][foo]"));
     }
+
+    @Test
+    public void setTopLevelTagsWithMap() {
+        Event.setIllegalTagsAction(Event.IllegalTagsAction.RENAME.toString());
+        final Event event = new Event();
+        event.setField("[tags][foo]", "bar");
+
+        assertEquals(event.getField(Event.TAGS), Collections.singletonList(Event.TAGS_FAILURE_TAG));
+        assertEquals(event.getField("[_tags][foo]"), "bar");
+    }
+
+    @Test
+    public void createEventWithTopLevelTagsWithMap() {
+        Event.setIllegalTagsAction(Event.IllegalTagsAction.RENAME.toString());
+        Map<String, Object> inner = new HashMap<>();
+        inner.put("poison", "true");
+        Map<String, Object> data = new HashMap<>();
+        data.put("tags", inner);
+        final Event event = new Event(data);
+
+        assertEquals(event.getField(Event.TAGS), Collections.singletonList(Event.TAGS_FAILURE_TAG));
+        assertEquals(event.getField("[_tags][poison]"), "true");
+    }
+
+
+    @Test
+    public void allowTopLevelTagsWithMap() {
+        Event.setIllegalTagsAction(Event.IllegalTagsAction.WARN.toString());
+        final Event event = new Event();
+        event.setField("[tags][foo]", "bar");
+        Event.setIllegalTagsAction(Event.IllegalTagsAction.RENAME.toString());
+
+        assertNull(event.getField(Event.TAGS_FAILURE_FIELD));
+        assertEquals(event.getField("[tags][foo]"), "bar");
+    }
+
+    @Test
+    public void allowCreatingEventWithTopLevelTagsWithMap() {
+        Event.setIllegalTagsAction(Event.IllegalTagsAction.WARN.toString());
+        Map<String, Object> inner = new HashMap<>();
+        inner.put("poison", "true");
+        Map<String, Object> data = new HashMap<>();
+        data.put("tags", inner);
+        final Event event = new Event(data);
+        Event.setIllegalTagsAction(Event.IllegalTagsAction.RENAME.toString());
+
+        assertNull(event.getField(Event.TAGS_FAILURE_FIELD));
+        assertEquals(event.getField("[tags][poison]"), "true");
+    }
```

Comment:
A helper could make managing which mode was active simpler throughout the tests, since they wouldn't have to worry about restoring the previous value:

```suggestion
    }
    
    public void withIllegalTagsAction(final Event.IllegalTagsAction temporaryIllegalTagsAction, final Runnable runnable) {
        try {
            final IllegalTagsAction previous = Event.ILLEGAL_TAGS_ACTION;
            Event.setIllegalTagsAction(temporaryIllegalTagsAction.toString());
            
            runnable.run();
        } finally {
            Event.setIllegalTagsAction(temporaryIllegalTagsAction.toString());
        }
    }
```

Usage:

~~~ java
    @Test
    void descriptiveTestName() {
        withIllegalTagsAction(Event.IllegalTagsAction.WARN), () -> {
            // your test code here
        });
    }
~~~